_posts/2018-03-19-from-python-to-go-to-rust.markdown

```diff
@@ -0,0 +1,80 @@
+
+I had the famous feeling "when it compiles it works", and once you've experienced that, __there is no way back__. You try stuff, you follow the friendly compiler error messages, you fix things, and then _tada_, it works!
+
+Ok so, at this point I knew what I wanted from the "perfect" backend language:
+
+1. __Static__ and __strong__ typing
+2. Most of the stuff checked at __compile time__ (please, no exceptions!)
+3. __No `null`__
+4. __No mutability__
+5. Handle __concurrency__ nicely
+
+I see you coming: "hey, this is __Haskell__"! Yeah, I may write another blog post about it, but for me the problem with Haskell is not the language itself, it's what goes with it: elit mindset from the outside, poor documentation and little to no efforts to make it accessible for beginners.
+
+"Hey, and what about Scala?!". What do you mean by Scala? The better Java? The functional programming language with scalaz? The Object Orienting Programming Functional language that may or may not fail at runtime with a `java.lang.NullPointerException` and needs a 4GB JVM running? I tried it some years ago and definitievely, this is a no go for me.
+
+After discussing with a few people, I decided to give __Go__ a try. It has a __compiler__, __no exceptions__, no `null` (but __null values__) and can handle __concurrency__ nicely.
+
+## Into Go,
+
+I decided to rewrite an internal project that was already done in Python using Go. Just to get a feeling of the differences between the two.
+
+First feeling: learning __Go__ was so easy. In __one evening__, I was able to compile a Proof Of Concept of the projet with basic features developped and some tests written. This was a very pleasant feeling, I was adding features very fast. The compiler messages were helpful, everything was fine.
+
+And at some point, the tragedy started. I needed to add a field to some struct, so I just modified the struct and was ready to analyze the compiler messages to know where this struct was used in order to add the field where it was needed.
+
+I compiled the code and … no error message. Everything went fine. But?! I just added a field to a struct, the compiler should say that my code is not good anymore because I'm not initializing the value where it should be!
+
+The problem is that, not providing a value to a struct is not a problem in __Go__. This value will default to it's [zero value](https://tour.golang.org/basics/12) and everything will compile. This was the __show stopper__ for me. I realized that I __couldn't rely on the compiler__ to get my back when I was doing mistakes. At this point, I was wondering: why should I bother learning __Go__ if the compiler can't do much better than [Python and mypy](http://mypy-lang.org/)? Of course concurrency is much better with __Go__, but the downside of not being able to rely on the compiler was too much for me.
+
```

Comment:
My 2 cents here and a personal opinion: Go is much safer, scalable, performant and readable than python. Minimizing Go's advantage over Python to it's concurrency feature is little misleading imho.

Reply:
> I __couldn't rely on the compiler__ to get my back when I was doing mistakes [...] if the compiler can't do much better than [Python and mypy](http://mypy-lang.org/)?

Again, my personal view on this: If you're comparing the Go toolchain with the rust compiler, then I agree it's less strict. Or rather said: the language definition is less restrictive and thus the compiler can not detect as many potential runtime failures as the Rust one.
But the Go native toolchain (not only its compiler) is several orders of magnitude better and safer than Python. IMHO it __does__ much better than Python although it is not as strict as the Rust compiler.

Reply:
And don't get me wrong, I coincidentally started to work with Rust a few days ago and I enjoy it as well. But putting Python and Go toolchains at almost the same level is a bit unfair to me.

However your Go vs Rust comparison makes sense to me although it's lacking a few data points. A good article showing how much stricter and safer Rust is:
https://medium.com/@deckarep/paradigms-of-rust-for-the-go-developer-210f67cd6a29